src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp

```diff
@@ -3952,12 +3952,41 @@ CTranslatorDXLToPlStmt::TranslateDXLDirectDispatchInfo
 	{
 		return NIL;
 	}
-	
+
 	CDXLDatumArray *dxl_datum_array = (*dispatch_identifier_datum_arrays)[0];
 	GPOS_ASSERT(0 < dxl_datum_array->Size());
-		
-	ULONG hash_code = GetDXLDatumGPDBHash(dxl_datum_array);
+
 	const ULONG length = dispatch_identifier_datum_arrays->Size();
+
+	if (dxl_direct_dispatch_info->FContainsRawValues()) {
+		List *segids_list = NIL;
+		Const *const_expr = NULL;
+
+		for (ULONG ul = 0; ul < length; ul++)
+		{
+			CDXLDatumArray *dispatch_identifier_datum_array = (*dispatch_identifier_datum_arrays)[ul];
+			GPOS_ASSERT(1 == dispatch_identifier_datum_array->Size());
+			const_expr = (Const *) m_translator_dxl_to_scalar->TranslateDXLDatumToScalar(
+				(*dispatch_identifier_datum_array)[0]
+			);
+
+			if ((const_expr->constvalue >= -1 && const_expr->constvalue < (INT)m_num_of_segments))
```

Comment:
```suggestion
			if ((const_expr->constvalue >= -1 && DatumGetInt32(const_expr->constvalue) < (INT)m_num_of_segments))
```

The higher bits of `Datum` cannot be assumed to be all zero, we should explicitly extract the int out of the datum here.